### PR TITLE
[Examples/Docs] Tic-tac-toe

### PR DESCRIPTION
**Pull request by amnn**
## Description 

A README for the tic-tac-toe app example, derived from:

https://github.com/MystenLabs/multisig_tic-tac-toe/blob/main/README.md

And updates to the docs for the tic-tac-toe app example to reflect changes to
modernise the guide and bring all its source code into the `sui`
mono-repo.

This is the last major example that lived in the `sui_programmability` directory.

## Test plan 

:eyes:

## Stack

- #18525 
- #18526 
- #18557 

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: docs/content/guides/developer/app-examples/tic-tac-toe.mdx [hunk 1]
```diff
@@ -2,354 +2,336 @@
 title: Tic-Tac-Toe
 ---
 
-This guide covers three different implementations of the game tic-tac-toe on Sui. The first example utilizes a centralized admin that marks the board on the users’ behalf. The second example utilizes a shared object that both users can mutate. And the third example utilizes a multisig, where instead of sharing the game board, it's in a 1-of-2 multisig of both users’ accounts. This guide compares and contrasts the design philosophies behind the three different games, as well as the pros and cons of each.
+This guide covers three different implementations of the game tic-tac-toe on Sui. The first example utilizes a centralized admin that marks the board on the users’ behalf. The second example utilizes a shared object that both users can mutate. And the third example utilizes a multisig, where instead of sharing the game board, it's in a 1-of-2 multisig of both users’ accounts.
 
-## tic_tac_toe.move
+## owned.move
 
 In this first example of tic-tac-toe, the game object, including the game board, is controlled by a game admin.
```
Comment (ronny-mysten) [inline docs/content/guides/developer/app-examples/tic-tac-toe.mdx:10]: Any reason not to use the plugin?
{@inject: examples/tic-tac-toe/move/sources/owned.move#struct=Game noComments}

File: docs/content/guides/developer/app-examples/tic-tac-toe.mdx [hunk 2]
```diff
@@ -2,354 +2,336 @@
...
     player: address,
-    row: u64,
-    col: u64,
+    row: u8,
+    col: u8,
 }
 ```
-The main logic of the game is in the following `create_game` function.
+
+Games are created with the `new` function:
```
Comment (ronny-mysten) [inline docs/content/guides/developer/app-examples/tic-tac-toe.mdx:34]: {@inject: examples/tic-tac-toe/move/sources/owned.move#fun=new noComments}

I can do these on this PR or a separate one.

File: examples/tic-tac-toe/README.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
+# Tic tac toe
+
+This is an end-to-end example for on-chain tic-tac-toe. It includes:
+
+- A [Move package](./move), containing two protocols for running a game of
+  tic-tac-toe. One that uses shared objects and consensus and another
+  that uses owned objects, and the fast path.
```
Comment (ronny-mysten) [inline examples/tic-tac-toe/README.md:7]: Maybe define "fast path" in a couple of words?

File: examples/tic-tac-toe/README.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+- A [React front-end](./ui), in TypeScript built on top of
+  `create-react-dapp`, using the TS SDK and `dapp-kit`.
+- A [Rust CLI](./cli), using the Rust SDK.
+- [Scripts](./scripts) to publish packages and update configs used
+  while building the front-end and CLI.
+
+## Shared tic tac toe
+
+In the shared protocol, player X creates the `Game` as a shared object
+and players take it in turns to place marks. Once the final move is
```
Comment (ronny-mysten) [inline examples/tic-tac-toe/README.md:17]:
```suggestion
and players take turns to place marks. Once the final move is
```

File: examples/tic-tac-toe/README.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    Admin->>Game: ...
+    Player O->>Game: ...
+    activate Admin
+    Admin->>Game: place_mark
+    Admin->>Player O: [Trophy]
+    deactivate Admin
+    Admin->>Game: burn
+```
+
+## Multi-sig tic-tac-toe
```
Comment (ronny-mysten) [inline examples/tic-tac-toe/README.md:85]:
```suggestion
## Multisig tic-tac-toe
```
It's multisig in the doc above (and in docs in general). It's consistent here so tomato, tomato.

File: examples/tic-tac-toe/README.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+of the multi-sig account to `burn` the game. As the player is the
+sponsor, they will receive the storage rebate for performing the
+clean-up.
+
+The multi-sig account does not own anything other than the game object
+(it does not have any gas coins of its own), so the player sponsors
+the transaction, using one of its own gas coins.
+
+Sharing a resource while avoiding consensus by transfering it to a
+multi-sig account can be generalised from two accounts to up to ten
```
Comment (ronny-mysten) [inline examples/tic-tac-toe/README.md:113]:
```suggestion
multi-sig account can be generalized from two accounts to a max of ten
```
"to up to" just reads odd to me. Also, 'merica!

File: examples/tic-tac-toe/README.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+addresses.
+
+## Pros and cons
+
+The shared protocol's main benefit is that its on-chain logic and
+client integration are straightforward, and its main downside is that
+it relies on consensus for ordering.
+
+In contrast, the owned protocol uses only fast-path transactions (no
+consensus), but its on-chain logic is more complicated because it
```
Comment (ronny-mysten) [inline examples/tic-tac-toe/README.md:130]: Maybe use this parenthetical in the first note (no consensus) and then not needed here.

File: examples/tic-tac-toe/README.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+needs to manage the `TurnCap`, and its off-chain logic is complicated
+either by a third party service to act as an Admin, or a multi-sig and
+sponsored transaction set-up. When using multi-sig, care also needs to
```
Comment (ronny-mysten) [inline examples/tic-tac-toe/README.md:133]:
```suggestion
sponsored transaction setup. When using multi-sig, care also needs to
```
nit

File: docs/content/guides/developer/app-examples/tic-tac-toe.mdx [hunk 9]
```diff
@@ -2,354 +2,336 @@
...
-    * &mut game.cur_turn = game.cur_turn + 1;
-    addr = game.get_cur_turn_address();
-    mark.during_turn = true;
-    transfer::transfer(mark, addr);
+  // ...
 }
 ```
 
-The first function is straightforward. The player sends the location of the mark to the multisig account. Then in the next function, the multisig actually places down the mark the player requested, as well as all the logic to check to see if there is a winner, end the game, and award a player a trophy if so, or to advance to the next player’s turn if not. See the [multisig_tic-tac-toe repo](https://github.com/MystenLabs/multisig_tic-tac-toe) for the full source code on this version of the game.
+([Source](https://github.com/MystenLabs/sui/blob/main/examples/tic-tac-toe/ui/src/pages/Game.tsx)).
```
Comment (ronny-mysten) [inline docs/content/guides/developer/app-examples/tic-tac-toe.mdx:331]: Could also use this construct to make the code available without user needing to navigate away:
```
<details>
   <summary>
      Toggle source code
   </summary>
   {@inject: examples/tic-tac-toe/ui/src/pages/Game.tsx}
</details>
```
This puts the relative path at the top, but it's not clickable.

**Review by ronny-mysten — APPROVED** [review]
Noted a few things, but nothing of note...

File: docs/content/guides/developer/app-examples/tic-tac-toe.mdx (hunk 2 above)
Comment (amnn) [reply docs/content/guides/developer/app-examples/tic-tac-toe.mdx:34]: Thanks @ronny-mysten ! I'd appreciate if you could show me how it's done this time around (in this PR is good). I wasn't sure what the status of the inject plugin was (whether it was ironed out, or whether there might be issues/edge cases), so I opted to focus on the content, but I'm happy for this to use it.

Couple of things that I thought might be edge cases:

- In one place, I reference only part of a function, with parts snipped out.
- I'm referencing code with a `.tsx` extension, which may complicate the logic of finding snippets.

Comment (ronny-mysten) [reply docs/content/guides/developer/app-examples/tic-tac-toe.mdx:36]: @amnn You're right. I added some features for move that aren't available for other langs yet (like removing comments). I also didn't consider cutting out more than one section at a time, but that should be a painless add. Anyway, I just updated the move parts for now.   

I left the copied code in there for now just so you can check it. That would obvs be deleted before publish. I also added the code collapsible for each `Source` link. I think both are valuable?

Comment (amnn) [reply docs/content/guides/developer/app-examples/tic-tac-toe.mdx:36]: Thanks very much @ronny-mysten ! And yes, I agree that both are valuable.

File: examples/tic-tac-toe/move/sources/owned.move
Comment (amnn) [inline examples/tic-tac-toe/move/sources/owned.move:1]: I'm not attached to the blank lines if they are annoying for the snippet tool -- we can get rid of them. I need to do it lower down the stack though (and republish etc).

File: docs/content/guides/developer/app-examples/tic-tac-toe.mdx (hunk 2 above)
Comment (amnn) [reply docs/content/guides/developer/app-examples/tic-tac-toe.mdx:36]: One thing we need to tweak is tab sizing -- JS/TS code snippets tend to use tabs, and for me they are showing up as tab width 4, but I think folks tend to use tab-width 2 when editing them. I think it's as simple as setting a style to control tab size -- I'll look into it.

**Comment by amnn** [comment]
@ronny-mysten, one final thing that it would be good to get your thoughts on is whether there's a way to fold the "toggle" UI for the full source code and the link to the full source code. I think we talked about this before and you mentioned that you can't include a link in the toggle summary because clicking on it will just activate the toggle, and not go to the link, but is there some other way to do it? Having the link to source hanging out at below the detail toggle feels kind of strange.

**Pull request merged**